pipeline.go

```diff
@@ -14,29 +14,76 @@ type Pipeline struct {
 	Steps []interface{} `yaml:"steps"`
 }
 
-func generateProjectSteps(step interface{}, projects []Project) []interface{} {
+func generateProjectSteps(steps []interface{}, step interface{}, projects []Project) []interface{} {
 	projectSteps := make([]interface{}, 0)
+
 	for _, project := range projects {
 		stepCopy := deepcopy.Copy(step)
 		stepCopyMap := stepCopy.(map[interface{}]interface{})
 
 		if project.checkProjectRules(stepCopyMap) {
+			// Unique project level label
 			stepCopyMap["label"] = fmt.Sprintf("%s %s", stepCopyMap["label"], project.Label)
+
+			// Set default buildpipe environment variables and
+			// set project env vars as step env vars
 			env := stepCopyMap["env"].(map[interface{}]interface{})
 			env["BUILDPIPE_PROJECT_LABEL"] = project.Label
 			env["BUILDPIPE_PROJECT_PATH"] = project.getMainPath()
-
 			for envVarName, envVarValue := range project.Env {
 				env[envVarName] = envVarValue
 			}
 
+			// Unique project level key, if present
+			if val, ok := stepCopyMap["key"]; ok {
+				stepCopyMap["key"] = fmt.Sprintf("%s %s", val, project.Label)
+			}
+
+			// If the step includes a depends_on clause, we need to validate whether each dependency
+			// is a project-scoped step. If so, the dependency has the current project name added
+			// to it to match the unique key given above.
+			if val, ok := stepCopyMap["depends_on"]; ok {
+				scopedDeps := make([]string, 0)
+				dependencyList := val.([]interface{})
+
+				for _, dependency := range dependencyList {
+					depStr := dependency.(string)
+					if isDependencyProjectScopeStep(steps, depStr) {
+						scopedDeps = append(scopedDeps, fmt.Sprintf("%s %s", depStr, project.Label))
+					} else {
+						scopedDeps = append(scopedDeps, depStr)
+					}
+				}
+
+				stepCopyMap["depends_on"] = scopedDeps
+			}
 			projectSteps = append(projectSteps, stepCopy)
 		}
 	}
 
 	return projectSteps
 }
 
+func isDependencyProjectScopeStep(steps []interface{}, dependencyName string) bool {
```

Comment:
```suggestion
func isProjectScopeStep(steps []interface{}, stepKey string) bool {
```
I don't think this function is necessarily limited to use with `depends_on` (although I suppose in practice it is, because it seems to be the only use case for `key` right now).

Reply:
Maybe split this up into `findStepByKey` and `isProjectStep`?